conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/ImmutableCollectionsModule.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.serialization;
+
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Values deserialized into {@link List}, {@link Set}, and {@link Map}, use Guava implementations of
+ * {@link ImmutableList}, {@link ImmutableSet}, and {@link ImmutableMap} respectively.
+ */
+final class ImmutableCollectionsModule extends SimpleModule {
+    private static final long serialVersionUID = 1L;
+
+    ImmutableCollectionsModule() {
+        super(ImmutableCollectionsModule.class.getCanonicalName());
+        addAbstractTypeMapping(List.class, ImmutableList.class);
+        addAbstractTypeMapping(Map.class, ImmutableMap.class);
+        addAbstractTypeMapping(Set.class, ImmutableSet.class);
+    }
+}
```

Comment:
This seems like a different change that should be done in a different PR.
Also, @iamdanfox raises a good point that if we want to standardize on these collections, why not make the change in conjure-java?
That way, users could also benefit from the added type-safety (it's obvious that these values are immutable, so never necessary to copy them).

Reply:
using interface types here gives us the freedom to move to another implementation without breaking all conjure consumers.
If we update conjure-java instead, we will break every generated service which consumes or return a list/set/map because the method signature will change.

Reply:
At the moment conjure-undertow generated code is the only place where our generated code depends on guava, and that's for a single `ImmutableList.of()` that I'm considering removing for simplicity.

Reply:
We can keep returning List etc in the interface then, and only use immutables in the implementation

Reply:
Then we'd need this change anyhow, otherwise we'd face the same problem for wrapper types used as the request/response body.